collect_app/src/main/java/org/odk/collect/android/formmanagement/metadata/FormMetadataParser.kt

```diff
@@ -57,75 +56,70 @@ object FormMetadataParser {
+            getFirstPrimaryInstanceGeopointXPath(geopointXPaths, repeatXPaths, mainInstanceRoot, null)
         }
     }
 
-    /**
-     * Returns the reference of the first geopoint in the body that is not in a repeat.
-     */
-    private fun getFirstToplevelBodyGeoPointXPath(model: Element, body: Element): String? {
-        for (elementId in 0 until body.childCount) {
-            val child = body.getElement(elementId) ?: continue
-            val ref = child.getAttributeValue(null, "ref")
-            if (child.name == "input" && isGeopoint(model, ref)) {
-                return ref
-            } else if (child.name == "group") {
-                return getFirstToplevelBodyGeoPointXPath(model, child)
+    private fun getGeopointXPaths(model: Element): List<String> {
+        val geopointXPaths = mutableListOf<String>()
+        for (elementId in 0 until model.childCount) {
+            val child = model.getElement(elementId) ?: continue
+            if (child.name == "bind" && child.getAttributeValue(null, "type") == "geopoint") {
+                geopointXPaths.add(child.getAttributeValue(null, "nodeset"))
             }
         }
-        return null
+        return geopointXPaths
     }
 
-    /**
-     * Returns the XPath path for the first geopoint in the primary instance that is before the given
-     * reference and not in a repeat.
-     */
-    private fun getInstanceGeoPointBeforeXPath(model: Element, mainInstanceRoot: Element): String? {
-        for (elementId in 0 until mainInstanceRoot.childCount) {
-            val child = mainInstanceRoot.getElement(elementId) ?: continue
-            val ref = "/${mainInstanceRoot.name}/${child.name}"
-            if (isGeopoint(model, ref)) {
-                return ref
+    private fun getRepeatXPaths(body: Element): List<String> {
+        val repeatXPaths = mutableListOf<String>()
+        for (elementId in 0 until body.childCount) {
+            val child = body.getElement(elementId) ?: continue
+            if (child.name == "repeat") {
+                repeatXPaths.add(child.getAttributeValue(null, "nodeset"))
             } else if (child.childCount > 0) {
-                return getInstanceGeoPointBeforeXPath(model, child)
+                repeatXPaths.addAll(getRepeatXPaths(child))
             }
         }
-        return null
+        return repeatXPaths
     }
 
-    private fun containsSetgeopointAction(model: Element): Boolean {
-        for (elementId in 0 until model.childCount) {
-            val child = model.getElement(elementId) ?: continue
-            if (child.name == SetGeopointActionHandler.ELEMENT_NAME) {
-                return true
+    private fun getFirstPrimaryInstanceGeopointXPath(
+        geopointXPaths: List<String>,
+        repeatXPaths: List<String>,
+        mainInstanceRoot: Element,
+        parentXPath: String?
+    ): String? {
+        for (elementId in 0 until mainInstanceRoot.childCount) {
```

Comment:
There are so many different kinds of ids everywhere that I think `position` would be easier to understand than `elementId` here.

Reply:
Done.